src/client.rs

```diff
@@ -285,71 +280,181 @@ impl Client {
-            &self.server_url,
-            &self.bucket_name,
-            &self.collection_name,
-        )?;
+        let remote_timestamp = match expected.into() {
+            Some(v) => v,
+            None => {
+                debug!("Obtain current timestamp.");
+                get_latest_change_timestamp(
+                    &self.server_url,
+                    &self.bucket_name,
+                    &self.collection_name,
+                )?
+            }
+        };
 
-        if let Some(collection) = stored {
+        if let Some(ref collection) = stored {
             let up_to_date = collection.timestamp == remote_timestamp;
             if up_to_date && self.verifier.verify(&collection).is_ok() {
                 debug!("Local data is up-to-date and valid.");
-                return Ok(collection.records);
+                return Ok(collection.to_owned());
             }
         }
 
         info!("Local data is empty, outdated, or has been tampered. Fetch from server.");
-        let collection = self._fetch_records_and_verify(remote_timestamp)?;
+        let (local_records, local_timestamp) = match stored {
+            Some(c) => (c.records, Some(c.timestamp)),
+            None => (Vec::new(), None),
+        };
+
+        let changeset = get_changeset(
+            &self.server_url,
+            &self.bucket_name,
+            &self.collection_name,
+            Some(remote_timestamp),
+            local_timestamp,
+        )?;
+
+        debug!(
+            "Apply {} changes to {} local records",
+            changeset.changes.len(),
+            local_records.len()
+        );
+        let merged = merge_changes(local_records.to_vec(), changeset.changes);
+
+        let collection = Collection {
+            bid: self.bucket_name.to_owned(),
+            cid: self.collection_name.to_owned(),
+            metadata: changeset.metadata,
+            records: merged,
+            timestamp: changeset.timestamp,
+        };
+
+        debug!("Verify signature after merge of changes with previous local data.");
+        self.verifier.verify(&collection)?;
+
+        debug!("Store collection with key={:?}", storage_key);
         let collection_bytes: Vec<u8> = serde_json::to_string(&collection)?.into();
         self.storage.store(&storage_key, collection_bytes)?;
-        Ok(collection.records)
+
+        Ok(collection)
+    }
+}
+
+fn merge_changes(
+    local_records: Vec<KintoObject>,
+    remote_changes: Vec<KintoObject>,
+) -> Vec<KintoObject> {
+    // Merge changes by record id and delete tombstones.
+    let mut local_by_id: HashMap<String, KintoObject> = HashMap::new();
+    for record in local_records {
+        local_by_id.insert(record["id"].to_string(), record);
+    }
+    for change in remote_changes.iter().rev() {
+        let id = change["id"].to_string();
+        if change
+            .get("deleted")
+            .unwrap_or(&json!(false))
+            .as_bool()
+            .unwrap_or(false)
```

Comment:
Here's a version that provides the default value only once:
```suggestion
            .map(|v| v.as_bool())
            .flatten()
            .unwrap_or(false)
```
Not sure whether that's easier to read, though.